# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/SharedFontList.h
# Commit: 12fc56d7782c
# Full Hash: 12fc56d7782c1aed1b2975d14a3c9d9f4c04b3a3
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2020-04-08 21:42:38
# Description:
#   Bug 1627605 - Avoid creating strings with shared-memory buffers that could become unmapped on a font-list refresh. r=jwatt
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70025
# ==============================================================================

diff -r b40b0c09c1fe -r 12fc56d7782c gfx/thebes/SharedFontList.h
--- a/gfx/thebes/SharedFontList.h	Wed Apr 08 09:46:34 2020 +0000
+++ b/gfx/thebes/SharedFontList.h	Wed Apr 08 09:53:35 2020 +0000
@@ -92,9 +92,12 @@
 
   const nsCString AsString(FontList* aList) const {
     MOZ_ASSERT(!mPointer.IsNull());
-    nsCString res;
-    res.AssignLiteral(static_cast<const char*>(mPointer.ToPtr(aList)), mLength);
-    return res;
+    // It's tempting to use AssignLiteral here so that we get an nsCString that
+    // simply wraps the character data in the shmem block without needing to
+    // allocate or copy. But that's unsafe because in the event of font-list
+    // reinitalization, that shared memory will be unmapped; then any copy of
+    // the nsCString that may still be around will crash if accessed.
+    return nsCString(static_cast<const char*>(mPointer.ToPtr(aList)), mLength);
   }
 
   void Assign(const nsACString& aString, FontList* aList);